### PR TITLE
[Merged by Bors] - feat: add API for multiplicative characters

### PR DESCRIPTION
**Pull request by MichaelStollBayreuth**
This adds some API lemmas for multiplicative characters. We also remove a couple of porting notes (the previous proof lines work again).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean [hunk 1]
```diff
@@ -542,17 +533,47 @@ theorem IsQuadratic.pow_odd {χ : MulChar R R'} (hχ : χ.IsQuadratic) {n : ℕ}
...
-variable {R : Type u} [CommMonoid R] [Fintype R] {R' : Type v} [CommRing R']
+variable {R : Type*} [CommMonoid R] [Fintype R] {R' : Type*} [CommRing R']
 
 /-- The sum over all values of a nontrivial multiplicative character on a finite ring is zero
 (when the target is a domain). -/
 theorem IsNontrivial.sum_eq_zero [IsDomain R'] {χ : MulChar R R'}
     (hχ : χ.IsNontrivial) : ∑ a, χ a = 0 := by
   rcases hχ with ⟨b, hb⟩
   refine eq_zero_of_mul_eq_self_left hb ?_
-  -- POrting note: `map_mul` isn't applied
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean:545]: It is not clear to me what this note complains about, so I'm removing it.

**Review by loefflerd — COMMENTED** [review]
I left some minor comments. I feel `LegendreSymbol` isn't the best place for this though; we should move it to somewhere in the algebra hierarchy. 

Incidentally, this file imports `Mathlib.Algebra.EuclideanDomain.Instances` but seems to compile fine with that import removed; getting rid of that will make it much easier to relocate without introducing undesirable imports (= upsetting Yael).

**Comment by MichaelStollBayreuth** [comment]
> we should move it to somewhere in the algebra hierarchy.

I'm planning to move `MulCharacter` to `NumberTheory.MulChar.Basic` (and add more API in a new `Lemmas` file); see [here on Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2313392.20Multiplicative.20characters/near/441938998). I'm also going to move `GaussSum` directly under `NumberTheory` (the original development was motivated by quadratic reciprocity, but Gauss sums are more generally uesful) and eventually add `NumberTheory.JacobiSum` as well.

File: Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean [hunk 2]
```diff
@@ -141,6 +132,9 @@ theorem map_nonunit (χ : MulChar R R') {a : R} (ha : ¬IsUnit a) : χ a = 0 :=
   χ.map_nonunit' a ha
 #align mul_char.map_nonunit MulChar.map_nonunit
 
+lemma map_unit (χ : MulChar R R') (a : Rˣ) : IsUnit (χ a) :=
+  IsUnit.map χ <| a.isUnit
```
[the remark this answers is not in the document]
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean:136]: It is sometimes convenient, though.

**Comment by loefflerd** [comment]
> Gauss sums are more generally useful

I agree strongly! So you're saying this move will follow in a future PR?

File: Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (loefflerd) [reply Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean:136]: Can you give a use case for this lemma where other existing lemmas, e.g. `IsUnit.map`, don't already achieve the same thing with equivalent or shorter code? I couldn't find one.

[the remark this answers is not in the document]
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean:136]: OK, I guess you are right. In my use case, I actually need the unit `χ a`, which is `equivToUnitHom χ a`, and so I can easily get rid of this lemma.

File: Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean [hunk 3]
```diff
@@ -532,17 +533,47 @@ theorem IsQuadratic.pow_odd {χ : MulChar R R'} (hχ : χ.IsQuadratic) {n : ℕ}
...
+variable {R : Type*} [CommMonoidWithZero R]
+
+/-- A multiplicative character on a finite commutative monoid has finite (= positive) order. -/
+lemma orderOf_pos (χ : MulChar M R) : 0 < orderOf χ := by
+  let e := MulChar.mulEquivToUnitHom (R := M) (R' := R)
+  rw [← MulEquiv.orderOf_eq e χ]
+  have : orderOf (e χ) ∣ Fintype.card Mˣ := by
+    refine orderOf_dvd_of_pow_eq_one ?_
+    ext1 x
+    change (e χ x) ^ _ = _
```
[the remark this answers is not in the document]
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LegendreSymbol/MulCharacter.lean]: I've added this as a `simp` lemma now.

**Comment by MichaelStollBayreuth** [comment]
> this file imports `Mathlib.Algebra.EuclideanDomain.Instances`

I have removed the import. Thanks for spotting this!

**Comment by loefflerd** [comment]
LGTM! 

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**